File: core/src/main/java/com/github/xsi640/qinquery/core/ast/OperatorExpression.java
package com.github.xsi640.qinquery.core.ast;

/**
 * @author SuYang
 */
public final class OperatorExpression implements Expression {

    private final Operator operator;

    public OperatorExpression(Operator operator) {
        this.operator = operator;
    }

    public Operator getOperator() {
        return operator;
    }

    @Override
    public <R, C> R accept(Visitor<R, C> visitor, C context) {
        return visitor.visit(this, context);
    }

    @Override
    public int priority() {
        return operator.priority;
    }
}

File: query-plus-core/src/main/kotlin/com/github/xsi640/queryplus/core/ast/OperatorExpression.kt
package com.github.xsi640.queryplus.core.ast

import com.github.xsi640.queryplus.core.ast.enums.Operator
import com.github.xsi640.queryplus.core.visitor.Visitor

class OperatorExpression(
    val left: Expression,
    val operator: String,
    val right: List<Expression>?
) : Expression {
    override fun <C, R> visit(context: C, visitor: Visitor<C, R>): R {
        return visitor.onOperator(context, this)
    }

    override fun priority(): Int {
        val op = Operator.of(this.operator)
        if (op != null)
            return op.priority
        return 200
    }
}